modules/core/src/main/scala/com.snowplowanalytics.iglu/client/validator/CirceValidator.scala

```diff
@@ -240,4 +246,65 @@ object CirceValidator extends Validator[Json] {
 
   private def fromValidationMessage(m: ValidationMessage): ValidatorReport =
     ValidatorReport(m.getMessage, m.getPath.some, m.getArguments.toList, m.getType.some)
+
+  type InitValidatorCache[F[_]] =
+    CreateLruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+  def validatorF[F[_]: Monad: InitValidatorCache]: F[ValidatorF[F, (SchemaKey, Int), Json]] =
+    CreateLruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+      .create(100)
+      .map(validatorFImpl(_))
+
+  def validatorFImpl[F[_]: Monad](
+    lookupJsonSchema: LruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+  ): ValidatorF[F, (SchemaKey, Int), Json] =
+    new ValidatorF[F, (SchemaKey, Int), Json] {
+
+      def lookupOrEvaluate(
+        key: Option[(SchemaKey, Int)],
+        schema: Json
+      ): F[Either[ValidatorError.InvalidSchema, JsonSchema]] = {
+        key
+          .fold(Monad[F].pure(none[Either[ValidatorError.InvalidSchema, JsonSchema]]))(
+            lookupJsonSchema.get(_)
+          )
+          .flatMap {
+            case Some(result) => result.pure[F]
+            case None =>
+              Either
+                .catchNonFatal(
+                  IgluMetaschemaFactory.getSchema(circeToJackson(schema), SchemaValidatorsConfig)
+                )
+                .leftMap(issue =>
+                  ValidatorError
+                    .InvalidSchema(
+                      NonEmptyList.of(ValidatorError.SchemaIssue("$", issue.getMessage))
+                    )
+                )
+                .pure[F]
+                .flatTap { result =>
+                  key.map(lookupJsonSchema.put(_, result)).sequence_
+                }
+          }
+      }
+
+      def validateF(
+        key: Option[(SchemaKey, Int)],
+        data: Json,
+        schema: Json
+      ): F[Either[ValidatorError, Unit]] =
+        lookupOrEvaluate(key, schema)
+          .map(_.flatMap { jsonschema =>
+            validateOnReadySchema(jsonschema, data).leftMap(ValidatorError.InvalidData.apply)
+          })
+
+      def checkSchemaF(
+        key: Option[(SchemaKey, Int)],
+        schema: Json
+      ): F[List[ValidatorError.SchemaIssue]] =
+        lookupOrEvaluate(key, schema)
+          .map {
+            case Right(_)      => Nil
```

Comment:
Previously, `checkSchema` called sth like `V4Schema.validate(json)`, it's not used here currently. Is it intentional?

Reply:
Good point, this needs more investigation.  In this implementation I am assuming that `lookupOrEvaluate` returns a `JsonSchema` only when the schema is valid against the v4 schema.  But I haven't fully checked if that's a valid assumption.  One solution could be to add a schema validation check into the `lookupOrEvaluate` function.

Reply:
I did some tests and it seems: `IgluMetaschemaFactory.getSchema` doesn't do any validations. So you could provide invalid schema (according to V4) and you would still get JsonSchema instance. And you could validate data against it! So to preserve current behaviour of client, we still need to validate schema against V4 explicitly